uaclient/api/api.py

```diff
@@ -35,8 +39,76 @@
 ]
 
 
+def _process_options(
+    options: List[str], fields: List[str]
+) -> Tuple[Dict[str, Any], List[ErrorWarningObject]]:
+    kwargs = {}
+    warnings = []
+
+    for option in options:
+        try:
+            k, v = option.split("=")
+        except ValueError:
+            raise APIError(
+                msg=API_BAD_ARGS_FORMAT.format(arg=option).msg,
+                msg_code=API_BAD_ARGS_FORMAT.name,
+            )
+
+        if not k or not v:
+            raise APIError(
+                msg=API_BAD_ARGS_FORMAT.format(arg=option).msg,
+                msg_code=API_BAD_ARGS_FORMAT.name,
+            )
+
+        if k not in fields:
+            warnings.append(
+                ErrorWarningObject(
+                    title=API_UNKNOWN_ARG.format(arg=k).msg,
+                    code=API_UNKNOWN_ARG.name,
+                    meta={},
+                )
+            )
+
+        kwargs[k] = v
+
+    return kwargs, warnings
+
+
+def _process_data(
+    data: str, fields: List[str]
+) -> Tuple[Dict[str, Any], List[ErrorWarningObject]]:
+    kwargs = {}
+    warnings = []
+
+    try:
+        json_data = json.loads(data)
+    except json.decoder.JSONDecodeError:
+        msg = API_JSON_DATA_FORMAT_ERROR.format(data=data)
+        raise APIError(msg=msg.msg, msg_code=msg.name)
+
+    for k, v in json_data.items():
+        if not k or not v:
+            raise APIError(
+                msg=API_BAD_ARGS_FORMAT.format(arg="{}:{}".format(k, v)).msg,
+                msg_code=API_BAD_ARGS_FORMAT.name,
+            )
+
+        if k not in fields:
+            warnings.append(
+                ErrorWarningObject(
+                    title=API_UNKNOWN_ARG.format(arg=k).msg,
+                    code=API_UNKNOWN_ARG.name,
+                    meta={},
+                )
+            )
+
+        kwargs[k] = v
+
+    return kwargs, warnings
```

Comment:
Some of this can be replaced by `endpoint.options_cls.from_dict(json_data)` and that would give us type checking as well

Reply:
We already perform that conversion just after we call that function. Although I do agree that maybe we just need to load the json data, we would then loose the warnings about extra keys. I don't have a strong opinion here, if we decide to simplify this, I don't see a problem